src/handler/mod.rs

```diff
@@ -0,0 +1,143 @@
+/// This module contains most of the low-level commands
+/// and grub variable modifications
+use anyhow::{bail, Result};
+use std::process::Command;
+use std::str;
+
+/// reboots the system if boot_counter is greater than 0 or can be forced too
+pub fn handle_reboot(force: bool) -> Result<()> {
+    if !force {
+        match get_boot_counter() {
+            Some(t) => {
+                if t <= 0 {
+                    bail!("countdown ended, check greenboot-rollback status")
+                }
+            }
+            None => bail!("boot_counter is not set"),
+        }
+    }
+    log::info!("restarting system");
+    let status = Command::new("systemctl").arg("reboot").status()?;
+    if status.success() {
+        return Ok(());
+    }
+    bail!("systemd returned error");
+}
+
+/// rollback to previous ostree deployment if boot counter is less than 0
+pub fn handle_rollback() -> Result<()> {
+    if let Some(t) = get_boot_counter() {
+        if t <= 0 {
+            log::info!("Greenboot will now attempt to rollback");
+            let status = Command::new("rpm-ostree").arg("rollback").status()?;
+            if status.success() {
+                return Ok(());
+            }
+            bail!(status.to_string());
+        }
+    }
+    log::info!("Rollback not initiated as boot_counter is either unset or not equal to 0");
```

Comment:
`Rollback not initiated, boot_counter {}` best to show the counter perhaps?

Reply:
Also, from my POV if the `log::info!` is not able to give us the specific info about what is happening:`either unset or not equal to 0 `, then we aren't getting all the information that we need.

Looking at a review that I left a month ago, from my POV `get_boot_counter` should return something other than `Option<i32>`. I said that maybe `Result<i32>` was the solution, but now I'm changing my opinion to `Result<Some<i32>>`. This way we should get all the info that we need for all the possible cases.

Reply:
You are also returning None when there is an error, so it makes no sense.

Reply:
it needs to return none as the execution should not fail at this point.
consider cases like: boot_counter=foo.

Reply:
In every place where the get_boot_counter function is used, the caller is returning `Result<()>`, so it is not true that the get_boot_counter funtion cannot fail.

Reply:
Lets break it if  I convey my idea clearly:

- Result<Option<i32>>  ->  We have to handle `Err, Ok(None, Some(t))` where `Err` and `None` are same, meaning: when a caller calls this function it has handle `Err` and `None` with same logic:  greenboot will want to set boot_counter 

- Option<i32> ->  So if it returns `None`, `Some(t)` , caller doesn't handle redundant logic. (basically handling the redundant logic in get_boot_counter itself)

Since get_boot_counter  is called multiple times  the previous step will need more  logic handling .

Reply:
Even if you want to handle `Err` and `None` with the same meaning, that should be the responsibility of the caller function, the callee should not obscure any information that it can return.

`get_boot_counter` can fail in the first line because something weird happened with the process that called`grub2-editenv`, and even if you are logging that info (more context needed in that error message btw), you are returning  `None`.
How can we distinguish between the `None` of "the process calling grub2-editenv failed" (maybe because of some selinux denial) and the `None` of  "boot counter not  set" if you are returning the same value in those two complete opposite cases?

Plus, you are also returning `None` when the boot_counter is set, but there is not a valid integer. 

Why is more valuable to obscure information rather than propagate it and let the callers handle it as they see fit?